autoload/maktaba/json.vim

```diff
@@ -0,0 +1,211 @@
+  let s:FALSE = maktaba#json#FALSE
+  lockvar! maktaba#json#NULL maktaba#json#TRUE maktaba#json#FALSE
+endif
+
+function! s:Ellipsize(str, limit) abort
+  return len(a:str) > a:limit ? a:str[ : a:limit - 4] . '...' : a:str
+endfunction
+
+function! s:FormatKeyAndValue(key, value) abort
+  if maktaba#value#IsString(a:key)
+    return printf('"%s": %s', a:key, maktaba#json#Format(a:value))
+  endif
+  throw maktaba#error#BadValue(
+      \ 'Non-string keys not allowed for JSON dicts: %s.', string(a:key))
+endfunction
+
+""
+" Formats {value} as a JSON text.
+" {value} may be any Vim value other than a Funcref or non-finite Float (or a
+" Dictionary or List containing either).
+" @throws BadValue if the input cannot be represented as JSON.
+function! maktaba#json#Format(value) abort
+  if a:value is s:NULL
+    return 'null'
+  elseif a:value is s:TRUE
+    return 'true'
+  elseif a:value is s:FALSE
+    return 'false'
+  elseif maktaba#value#IsNumeric(a:value)
+    let l:json = string(a:value)
+    if index(['nan', 'inf', '-nan', '-inf'], l:json) != -1
+      throw maktaba#error#BadValue(
+          \ 'Value cannot be represented as JSON: %s', l:json)
+    endif
+    return l:json
+  elseif maktaba#value#IsString(a:value)
+    let l:escaped = substitute(escape(a:value, '"\'), "\n", '\\n', 'g')
+    let l:escaped = substitute(l:escaped, "\t", '\\t', 'g')
+    " TODO(dbarnett): Escape other special characters.
+    return '"' . l:escaped . '"'
+  elseif maktaba#value#IsList(a:value)
+    let l:json_items = map(copy(a:value), 'maktaba#json#Format(v:val)')
+    return '[' . join(l:json_items, ', ') . ']'
+  elseif maktaba#value#IsDict(a:value)
+    if maktaba#value#IsCallable(a:value)
+      throw maktaba#error#BadValue(
+          \ 'Funcdict for func %s cannot be represented as JSON.',
+          \ string(a:value.func))
+    endif
+    let l:json_items =
+        \ map(items(a:value), 's:FormatKeyAndValue(v:val[0], v:val[1])')
+    return '{' . join(l:json_items, ', ') . '}'
+  endif
+  throw maktaba#error#BadValue(a:value)
+endfunction
+
+function! s:Consume(str, count) abort
+  return maktaba#string#StripLeading(a:str[a:count : ])
+endfunction
+
+function! s:ParsePartial(json, custom_values) abort
+  " null, true, or false
+  if a:json =~# '\m^null\>'
+    let l:value = get(a:custom_values, 'null', s:NULL)
+    return [l:value, s:Consume(a:json, 4)]
+  endif
+  if a:json =~# '\m^true\>'
+    let l:value = get(a:custom_values, 'true', s:TRUE)
+    return [l:value, s:Consume(a:json, 4)]
+  endif
+  if a:json =~# '\m^false\>'
+    let l:value = get(a:custom_values, 'false', s:FALSE)
+    return [l:value, s:Consume(a:json, 5)]
+  endif
+  " Number
+  " TODO(dbarnett): Handle scientific notation.
+  let l:num_match = matchstr(a:json, '\v^-?[0-9]+(\.[0-9]+)?>')
+  if !empty(l:num_match)
+    return [eval(l:num_match), s:Consume(a:json, len(l:num_match))]
+  endif
+  " String
+  let l:str_match = matchstr(a:json, '\v^"([^\\"]|\\.)*"')
```

Comment:
Eh, maybe. That is not valid by the JSON spec (nor are unquoted key strings), but since they're valid JS sloppy serializers tend to break those rules. See http://bugs.python.org/issue20298 for the analogous discussion for python's json library.

I think I'd be fine adding a `[nonstrict]` arg to `maktaba#json#Parse` to accept both of those forms as long as it's off by default, but there's a huge gotcha: the python implementation can't handle them, and I don't want a feature in the fallback unless we can get it in both implementations.